README.md
# docker-security-infrastructure
[![GitHub Stable Release](https://img.shields.io/badge/Release-4.1-blue.svg)](https://github.com/BradleyA/docker-security-infrastructure/releases/tag/4.1)
![GitHub Release Date](https://img.shields.io/github/release-date/BradleyA/docker-security-infrastructure?color=blue)
[![GitHub Commits Since](https://img.shields.io/github/commits-since/BradleyA/docker-security-infrastructure/4.1?color=orange)](https://github.com/BradleyA/docker-security-infrastructure/commits/)
[![GitHub Last Commits](https://img.shields.io/github/last-commit/BradleyA/docker-security-infrastructure.svg)](https://github.com/BradleyA/docker-security-infrastructure/commits/)

[![Open GitHub Issue](https://img.shields.io/badge/Open-Incident-brightgreen.svg)](https://github.com/BradleyA/docker-security-infrastructure/issues/new/choose)
[![GitHub Open Issues](https://img.shields.io/github/issues/BradleyA/docker-security-infrastructure?color=purple)](https://github.com/BradleyA/docker-security-infrastructure/issues?q=is%3Aopen+is%3Aissue)
[![GitHub Closed Issues](https://img.shields.io/github/issues-closed/BradleyA/docker-security-infrastructure?color=purple)](https://github.com/BradleyA/docker-security-infrastructure/issues?q=is%3Aclosed+is%3Aissue)
![GitHub pull requests](https://img.shields.io/github/issues-pr-raw/BradleyA/docker-security-infrastructure)
![GitHub pull requests](https://img.shields.io/github/issues-pr-closed-raw/BradleyA/docker-security-infrastructure)

[<img alt="GitHub Repo Clones" src="https://img.shields.io/static/v1?label=Repo_Clones&message=460&color=blueviolet">](https://github.com/BradleyA/docker-security-infrastructure/blob/master/images/clone.table.md)
[<img alt="GitHub Repo Views" src="https://img.shields.io/static/v1?label=Repo_Views&message=2851&color=blueviolet">](https://github.com/BradleyA/docker-security-infrastructure/blob/master/images/view.table.md)
[![GitHub Size](https://img.shields.io/github/repo-size/BradleyA/docker-security-infrastructure.svg)](https://github.com/BradleyA/docker-security-infrastructure/)
![Language Bash](https://img.shields.io/badge/%20Language-bash-blue.svg)
[![MIT License](http://img.shields.io/badge/License-MIT-blue.png)](LICENSE)

----

This repository contains shell scripts that check user ssh permissions, setup and manage both TLS and dockerd on Ubuntu 14.04 (upstart) and Ubuntu 16.04 (systemd).  Most scripts in this repository support --help and --usage options.
  
#### If you like this repository, select in the upper-right corner, [![GitHub stars](https://img.shields.io/github/stars/BradleyA/docker-security-infrastructure.svg?style=social&label=Star&maxAge=2592000)](https://GitHub.com/BradleyA/docker-security-infrastructure/stargazers/), thank you.

----

These bash scripts will create, copy, and check TLS public keys, private keys, and self-signed certificates for the docker user, daemon, and docker swarm. 
 * [docker-TLS](https://github.com/BradleyA/docker-security-infrastructure/tree/master/docker-TLS)
 
Setup a dockerd configuration file with dockerd flags for both systemd and Upstart.
 
 * [dockerd-configuration-options](https://github.com/BradleyA/docker-security-infrastructure/tree/master/dockerd-configuration-options)

These bash scripts allows users and administrators to make sure that the ssh files and directory permissions are correct.

 * [ssh](https://github.com/BradleyA/docker-security-infrastructure/tree/master/ssh)
 
[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

## Clone

To Install, change into a directory that you want to download the scripts. Use git to pull or clone these scripts into the directory. If you do not have Git installed then enter; "sudo apt-get install git" if using Debian/Ubuntu. Other Linux distribution install methods can be found here: https://git-scm.com/download/linux. On the GitHub page of this script use the "HTTPS clone URL" with the 'git clone' command.

     git clone https://github.com/BradleyA/docker-security-infrastructure
     cd docker-security-infrastructure

[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

#### Docker security links
 * [Docker Bench for Security](https://github.com/docker/docker-bench-security)
 * [Kubernetes Benchmark](https://github.com/aquasecurity/kube-bench)
 * [Center for Internet Security](https://www.cisecurity.org/)
 * [Notary](https://github.com/theupdateframework/notary)
 * [Clair](https://github.com/coreos/clair)
 * [Falco](https://sysdig.com/opensource/falco/)
 * [Cilium](https://github.com/cilium/cilium)
 * [Docker Security Tutorials](https://github.com/docker/labs/blob/master/security/README.md)
 * [Dagda](https://github.com/eliasgranderubio/dagda)
 * [Dockscan](https://github.com/kost/dockscan)
 * [Docker Enterprise Edition Compliance Controls](https://github.com/docker/compliance)
 * [Calico](https://www.projectcalico.org/)
 * [AppArmor](https://gitlab.com/apparmor/apparmor/wikis/home/)
 * [Hashicorp’s Vault](https://www.vaultproject.io/)
 * [OpenSCAP](https://www.open-scap.org/)
 * [Open Policy Agent](https://www.openpolicyagent.org/)
 * [REMnux](https://remnux.org/)
 * [SELinux](https://selinuxproject.org/page/Main_Page)
 * [Seccomp](https://www.kernel.org/doc/Documentation/prctl/seccomp_filter.txt)
 * [Anchore Engine](https://github.com/anchore/anchore-engine)
 * [AquaSec](https://www.aquasec.com/)
 * [Capsule8](https://capsule8.com/)
 * [Cavirin](https://cavirin.com/)
 * [Aporeto](https://www.aporeto.com/)
 * [Google Cloud Security Command Center](https://cloud.google.com/security-command-center/)
 * [Layered Insight](https://layeredinsight.com/)
 * [NeuVector](https://neuvector.com/)
 * [StackRox](https://www.stackrox.com/)
 * [Sysdig Secure](https://sysdig.com/products/secure/)
 * [Sysdig](https://sysdig.com/)
 * [Tenable - FlawCheck](https://www.tenable.com/products/tenable-io/container-security)

[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

#### To watch future updates in this repository select in the upper-right corner, the "Watch" list, and select Watching.

----

#### Contribute
Please do contribute!  Issues, comments, and pull requests are welcome.  Thank you for your help improving software.

[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

#### Author
[<img id="github" src="images/github.png" width="50" a="https://github.com/BradleyA/">](https://github.com/BradleyA/)    [<img src="images/linkedin.png" style="max-width:100%;" >](https://www.linkedin.com/in/bradleyhallen) [<img id="twitter" src="images/twitter.png" width="50" a="twitter.com/bradleyaustintx/">](https://twitter.com/bradleyaustintx/)       <a href="https://twitter.com/intent/follow?screen_name=bradleyaustintx"> <img src="https://img.shields.io/twitter/follow/bradleyaustintx.svg?label=Follow%20@bradleyaustintx" alt="Follow @bradleyaustintx" />    </a>         [![GitHub followers](https://img.shields.io/github/followers/BradleyA.svg?style=social&label=Follow&maxAge=2592000)](https://github.com/BradleyA?tab=followers)

[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

#### Stars
[![Stargazers repo roster for @BradleyA/docker-security-infrastructure](https://reporoster.com/stars/BradleyA/docker-security-infrastructure)](https://github.com/BradleyA/docker-security-infrastructure/stargazers)

[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

#### Forks
[![Forkers repo roster for @BradleyA/docker-security-infrastructure](https://reporoster.com/forks/BradleyA/docker-security-infrastructure)](https://github.com/BradleyA/docker-security-infrastructure/network/members)

[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

#### System OS script tested
 * Ubuntu 14.04.6 LTS (amd64,armv7l)
 * Ubuntu 16.04.7 LTS (amd64,armv7l)
 * Ubuntu 18.04.5 LTS (amd64,armv7l)
 * Raspbian GNU/Linux 10 (buster)

[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

#### Design Principles
 * Have a simple setup process and a minimal learning curve
 * Be usable as non-root - [some of these scripts require root]
 * Be easy to install and configure
 
[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)

#### License
MIT License

Copyright (c) 2020  [Bradley Allen](https://www.linkedin.com/in/bradleyhallen)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

[Return to top](https://github.com/BradleyA/docker-security-infrastructure/blob/master/README.md#docker-security-infrastructure)


CHANGELOG.md
**Notes:** Use the following example to guide you to; How to download an earlier release.

    git clone https://github.com/BradleyA/<REPOSITORY>.git --branch 4.1

## Next Version Under Construction {Version  4.2.????}  <img id="Construction" src="images/construction-icon.gif" width="120">
#### Release Name  prerelease 4.2.????
#### Release Date  ????
### Features Added
### Features Changes
### Features Deprecations
### Issues
 * close #11 dockerd-configuration-options - why systemd
 * close #37 Incident Report - dockerd-configuration-options/README - review and update BUG
 * close #72 docker-TLS/copy-host-2-remote-host-tls.sh - display_help description typos
 * close #73 Incident Report - dockerd-configuration-options/start-dockerd-with-systemd.sh - line 166: /bin/systemctl No such file or directory
 * close #74 Incident Report - docker-TLS/check-registry-tls.sh - remove environment variables

### Misc
* backlog
    * #10 dockerd-configuration-options/setup-dockerd.sh systemctl daemon-reload testing
    * #40 swarm/{docker-enter,docker-swarm-clear} - test swarm script from 2016 clusters and upgrade
    * #44 registry solve the incident requiring --disable-content-trust
    * #45 ufw/create-ufw.sh - create firewall Uncomplicated Firwwall (UFW)
    * #46 ufw/copy-ufw.sh - copy firewall Uncomplicated Firwwall (UFW)
    * #47 ufw/create-ufw.sh - Remove firewall Uncomplicated Firwwall (UFW)
    * #51 ssh/check-user-ssh.sh - check for additional authorized_keys when encrypted home directory
    * #58 ssh notes for design changes

# Latest Release -->  Version 4.1 
#### Release Name Production-4
#### Release Date 2019-12-30
* Majority of changes and updates for this release are in docker-security-infrastructure/docker-TLS

### Features Added
* docker-TLS/* 
  * Production standard 6.3.547 Architecture tree
  * Production standard 5.3.550 Copyright
  * Production standard 0.3.550 --help
  * Production standard 8.3.541 --usage
  * Production standard 4.3.550 Documentation Language
  * Production standard 1.3.550 DEBUG variable
  * added SA and FVT tests
  * update output for shellcheck incidents
  * improve output by adding color
* docker-TLS/check-ca-tls.sh 
  * created this script to test ca.pem close #49
  * update output and display_help
* docker-TLS/create-host-tls.sh 
  * add warning when ${CA_CERT} expires before cert.pem
  * removed hostname from key names in symbolic link
  * change file name standard to include site private CA date in all certs that are built from it
* docker-TLS/create-user-tls.sh add warning when ${CA_CERT} expires before user-cert.pem
* docker-TLS/copy-host-2-remote-host-tls.sh add default standard for Docker Swarm standalone manager
* docker-TLS/create-registry-tls.sh updated display_help, added cert duration dates & cert symbolic links

### Issues
* close #5 #48 docker-TLS/copy-user-2-remote-host-tls.sh docker-TLS/copy-host-2-remote-host-tls.sh changes for #5 #48 localhost does not use scp & ssh
* close #41 copy-registry-tls.sh - upgrade Production standard
* close #50 docker-TLS/check-host-tls.sh - modify output: add user message about cert expires 
* close #52 docker-TLS/c* - added production standard 8.0 --usage
* close #53 dockerd-configuration-options{setup-dockerd.sh,uninstall-dockerd-scripts.sh} - added production standard 8.0 --usage 
* close #54 ssh/check-user-ssh.sh change DEFAULT_USER_HOME 
* close #56 complete, release to production
* close #57 pre-commit add logging information to output / corrected incident with function not being call
* close #59 docker-TLS/check-ca-tls.sh - upgrade Production standard #59
   * docker-TLS/check-ca-tls.sh docker-TLS/check-registry-tls.sh #59 #61 updated Production standard 2.3.529 log format, 8.3.530 --usage, 1.3.531 DEBUG variable
* close #60 check-host-tls.sh - upgrade Production standard
* close #60 docker-TLS/create-site-private-public-tls.sh Production standard 2.3.529 log format, 8.3.530 --usage, 1.3.531 DEBUG variable
* close #61 docker-TLS/check-registry-tls.sh - upgrade Production standard
* close #62 docker-TLS/check-user-tls.sh - upgrade Production standard
* close #63 docker-TLS/copy-host-2-remote-host-tls.sh - upgrade Production standard
* close #64 docker-TLS/copy-user-2-remote-host-tls.sh - upgrade Production standard
* #64 docker-TLS/copy-user-2-remote-host-tls.sh Production standard 2.3.529 log format, 8.3.530 --usage, 1.3.531 DEBUG variable
* close #65docker-TLS/copy-host-2-remote-host-tls.sh  Production standard 2.3.529 log format, 8.3.530 --usage, 1.3.531 DEBUG variable
* close #66 docker-TLS/create-host-tls.sh upgrade Production standard
* close #67 docker-TLS/create-new-openssl.cnf-tls.sh Production standard 2.3.529 log format, 8.3.530 --usage, 1.3.531 DEBUG variable
* close #68 docker-TLS/create-registry-tls.sh - upgrade Production standard
* close #70 docker-TLS/create-user-tls.sh - upgrade Production standard

### Misc

## Version 3.234 
#### Release Name 
#### Release Date Apr 10, 2019
* Majority of changes and updates for this release are in docker-security-infrastructure/docker-TLS

### Features Added
* update EXAMPLES format
* update DOCUMENTATION change links
* update README
  * added registry examples
  * added Architecture tree
* new command/service
  * added docker-TLS/copy-registry-tls.sh #43
  * added docker-TLS/create-registry-tls.sh #41
  * added docker-TLS/create-registry-tls.sh #41
* update production standard 0.1.166 --help many updates to display_help
* added new user help hints messaging
* added production standard 3.0 shellcheck
* added production standard 6.1.177 ARCHITECTURE TREE section
* added production standard 7.0 Default variable value section


## Version  3.122
#### Release Name docker-security-infrastructure-scripts
#### Release Date Feb 20, 2019
* update README, added DEBUG, update display_help; update 1-5 productio…
* …ns standards


## Version  release.3.58
#### Release Name release.3.58
#### Release Date Sep 14, 2018

docker-TLS/check-ca-tls.sh
#!/bin/bash
# 	docker-TLS/check-ca-tls.sh  4.1.1211  2019-12-30T11:34:26.456665-06:00 (CST)  https://github.com/BradleyA/docker-security-infrastructure.git  uadmin  five-rpi3b.cptx86.com 3.565-13-g1455a67  
# 	   docker-TLS/*   New Release 4.1 
# 	docker-TLS/check-ca-tls.sh  3.563.1141  2019-12-22T20:15:30.690203-06:00 (CST)  https://github.com/BradleyA/docker-security-infrastructure.git  uadmin  five-rpi3b.cptx86.com 3.562 
# 	   docker-TLS/check-ca-tls.sh   created this script to test ca.pem  close #49 
# 	docker-TLS/check-ca-tls.sh  3.549.1117  2019-12-15T14:54:01.361686-06:00 (CST)  https://github.com/BradleyA/docker-security-infrastructure.git  uadmin  five-rpi3b.cptx86.com 3.548  
# 	   docker-TLS/check-ca-tls.sh   Production standard 5.3.550 Copyright  Production standard 0.3.550 --help  Production standard 4.3.550 Documentation Language  Production standard 1.3.550 DEBUG variable 
# 	docker-TLS/check-ca-tls.sh  3.543.1106  2019-12-13T16:20:51.645739-06:00 (CST)  https://github.com/BradleyA/docker-security-infrastructure.git  uadmin  five-rpi3b.cptx86.com 3.542  
# 	   Production standard 6.3.547  Architecture tree  Production standard 8.3.541 --usage 
# 	docker-TLS/check-ca-tls.sh  3.454.951  2019-10-13T15:24:56.576980-05:00 (CDT)  https://github.com/BradleyA/docker-security-infrastructure.git  uadmin  five-rpi3b.cptx86.com 3.453  
# 	   docker-TLS/check-ca-tls.sh docker-TLS/check-registry-tls.sh  #59 #61  updated Production standard 2.3.529 log format, 8.3.530 --usage, 1.3.531 DEBUG variable 
# 	docker-TLS/check-ca-tls.sh  3.293.760  2019-07-20T19:51:00.914284-05:00 (CDT)  https://github.com/BradleyA/docker-security-infrastructure  uadmin  six-rpi3b.cptx86.com 3.292  
# 	   complete, release to production close #56 
#86# docker-TLS/check-ca-tls.sh - shell script start and end dates of ${DEFAULT_CERT_DIR}/${DEFAULT_CA_CERT}
###  Production standard 3.0 shellcheck
###  Production standard 5.3.550 Copyright                                                  # 3.550
#    Copyright (c) 2020 Bradley Allen                                                       # 3.550
#    MIT License is online  https://github.com/BradleyA/user-files/blob/master/LICENSE      # 3.550
###  Production standard 1.3.550 DEBUG variable                                             # 3.550
#    Order of precedence: environment variable, default code
if [[ "${DEBUG}" == ""  ]] ; then DEBUG="0" ; fi   # 0 = debug off, 1 = debug on, 'export DEBUG=1', 'unset DEBUG' to unset environment variable (bash)
if [[ "${DEBUG}" == "2" ]] ; then set -x    ; fi   # Print trace of simple commands before they are executed
if [[ "${DEBUG}" == "3" ]] ; then set -v    ; fi   # Print shell input lines as they are read
if [[ "${DEBUG}" == "4" ]] ; then set -e    ; fi   # Exit immediately if non-zero exit status
if [[ "${DEBUG}" == "5" ]] ; then set -e -o pipefail ; fi   # Exit immediately if non-zero exit status and exit if any command in a pipeline errors
#
BOLD=$(tput -Txterm bold)
NORMAL=$(tput -Txterm sgr0)
YELLOW=$(tput setaf 3)
CYAN=$(tput   setaf 6)
WHITE=$(tput  setaf 7)

###  Production standard 7.0 Default variable value
DEFAULT_CERT_DIR="${HOME}/.docker"
DEFAULT_CA_CERT="ca.pem"

###  Production standard 8.3.541 --usage
COMMAND_NAME=$(echo "${0}" | sed 's/^.*\///')                                               # 3.541
display_usage() {
echo -e "\n${NORMAL}${COMMAND_NAME}\n   print start and end dates of ${DEFAULT_CERT_DIR}/${DEFAULT_CA_CERT}"
echo -e "\n${BOLD}USAGE${NORMAL}"
echo    "   ${YELLOW}Positional Arguments${NORMAL}"
echo    "   ${COMMAND_NAME} [<CERT_DIR>]"
echo -e "   ${COMMAND_NAME}  <CERT_DIR> [<CA_CERT>]\n"
echo    "   ${COMMAND_NAME} [--help | -help | help | -h | h | -?]"
echo    "   ${COMMAND_NAME} [--usage | -usage | -u]"
echo    "   ${COMMAND_NAME} [--version | -version | -v]"
}

###  Production standard 0.3.550 --help                                                     # 3.550
display_help() {
display_usage
#    Displaying help DESCRIPTION in English en_US.UTF-8, en.UTF-8, C.UTF-8                  # 3.550
echo -e "\n${BOLD}DESCRIPTION${NORMAL}"
echo    "Print start and end dates of a Docker CA, ${DEFAULT_CA_CERT}, in"
echo    "${DEFAULT_CERT_DIR}.  Other Docker CA file names in other directories"
echo    "can be checked using options."

###  Production standard 4.3.550 Documentation Language                                     # 3.550
#    Displaying help DESCRIPTION in French fr_CA.UTF-8, fr_FR.UTF-8, fr_CH.UTF-8
if [[ "${LANG}" == "fr_CA.UTF-8" ]] || [[ "${LANG}" == "fr_FR.UTF-8" ]] || [[ "${LANG}" == "fr_CH.UTF-8" ]] ; then
  echo -e "\n--> ${LANG}"
  echo    "<votre aide va ici>" # your help goes here
  echo    "Souhaitez-vous traduire la section description?" # Do you want to translate the description section?
elif ! [[ "${LANG}" == "en_US.UTF-8" ||  "${LANG}" == "en.UTF-8" || "${LANG}" == "C.UTF-8" ]] ; then  # 3.550
  new_message "${LINENO}" "${YELLOW}INFO${WHITE}" "  Your language, ${LANG}, is not supported.  Would you like to translate the description section?" 1>&2
fi

echo -e "\n${BOLD}ENVIRONMENT VARIABLES${NORMAL}"
echo    "If using the bash shell, enter; 'export DEBUG=1' on the command line to set"
echo    "the environment variable DEBUG to '1' (0 = debug off, 1 = debug on).  Use the"
echo    "command, 'unset DEBUG' to remove the exported information from the environment"
echo    "variable DEBUG.  You are on your own defining environment variables if"
echo    "you are using other shells."

###  Production standard 1.3.550 DEBUG variable                                             # 3.550
echo    "   DEBUG           (default off '0')  The DEBUG environment variable can be set"   # 3.550
echo    "                   to 0, 1, 2, 3, 4 or 5.  The setting '' or 0 will turn off"      # 3.550
echo    "                   all DEBUG messages during execution of this script.  The"       # 3.550
echo    "                   setting 1 will print all DEBUG messages during execution."      # 3.550
echo    "                   Setting 2 (set -x) will print a trace of simple commands"       # 3.550
echo    "                   before they are executed.  Setting 3 (set -v) will print"       # 3.550
echo    "                   shell input lines as they are read.  Setting 4 (set -e) will"   # 3.550
echo    "                   exit immediately if non-zero exit status is recieved with"      # 3.550
echo    "                   some exceptions.  Setting 5 (set -e -o pipefail) will do"       # 3.550
echo    "                   setting 4 and exit if any command in a pipeline errors.  For"   # 3.550
echo    "                   more information about the set options, see man bash."          # 3.550

echo    "   CERT_DIR        Certification directory (default '${DEFAULT_CERT_DIR}')"
echo    "   CA_CERT         Name of CA certificate (default '${DEFAULT_CA_CERT}')"

echo -e "\n${BOLD}OPTIONS${NORMAL}"
echo -e "Order of precedence: CLI options, environment variable, default code.\n"
echo    "   CERT_DIR        Certification directory (default '${DEFAULT_CERT_DIR}')"
echo    "   CA_CERT         Name of CA certificate (default '${DEFAULT_CA_CERT}')"

###  Production standard 6.3.547  Architecture tree
echo -e "\n${BOLD}ARCHITECTURE TREE${NORMAL}"  # STORAGE & CERTIFICATION
echo    "<USER_HOME>/                               <-- Location of user home directory"
echo    "└── <USER-1>/.docker/                      <-- User docker cert directory"
echo -e "    └── ca.pem                             <-- User tlscacert or symbolic link\n"
echo    "/etc/ "
echo    "└── docker/ "
echo    "    └── certs.d/                           <-- Host docker cert directory"
echo    "        ├── daemon/                        <-- Daemon cert directory"
echo    "        │   └── ca.pem                     <-- CA Cert"                            # 3.542
echo    "        └── <REGISTRY_HOST>:<REGISTRY_PORT>/ < Registry cert directory"
echo    "            └── ca.crt                     <-- Daemon registry domain cert"

echo -e "\n${BOLD}DOCUMENTATION${NORMAL}"
echo    "   https://github.com/BradleyA/docker-security-infrastructure/blob/master/docker-TLS/README.md"

echo -e "\n${BOLD}EXAMPLES${NORMAL}"
echo    "   User checking start and end dates of their Docker CA in \$HOME/.docker."
echo -e "\t${BOLD}${COMMAND_NAME}${NORMAL}\n"                                               # 3.550
echo    "   User checking start and end dates of their Docker CA in ."
echo -e "\t${BOLD}cd ~/.docker ; ${COMMAND_NAME} . ca.pem${NORMAL}\n"                       # 3.550
echo    "   To loop through a list of hosts a user could use, cluster-command.sh."          # 3.550
echo    "   An administrator may receive password and/or passphrase prompts from a"         # 3.550
echo    "   remote systen; running the following may stop the prompts."                     # 3.550
echo -e "\t${BOLD}ssh-copy-id <TLS_USER>@<REMOTE_HOST>${NORMAL}"                            # 3.550
echo    "   or"                                                                             # 3.550
echo -e "\t${BOLD}ssh-copy-id <TLS_USER>@<192.168.x.x>${NORMAL}"                            # 3.550
echo    "   If that does not resolve the prompting challenge then review man pages for"     # 3.550
echo    "   ssh-agent and ssh-add."                                                         # 3.550
echo    "   (https://github.com/BradleyA/Linux-admin/tree/master/cluster-command)"          # 3.550
echo -e "\t${BOLD}cluster-command.sh special '${COMMAND_NAME}'${NORMAL}\n"                  # 3.550
echo    "   Administrator checking start and end dates of another user by using:"
echo -e "\t${BOLD}sudo ${COMMAND_NAME} ~sam/.docker ca.pem${NORMAL}\n"                      # 3.550
echo    "   To loop through a list of hosts in a cluster an administrator could check the"
echo    "   start and end dates of Docker CA for user sam in /home/sam/.docker."
echo -e "\t${BOLD}cluster-command.sh special 'sudo ${COMMAND_NAME} /home/sam/.docker ca.pem'${NORMAL}\n"  # 3.550
echo    "   Administrator checking a Docker daemon CA by including sudo.  To use"
echo    "   ${COMMAND_NAME} on a remote hosts (one-rpi3b.cptx86.com) with ssh port"
echo    "   of 12323 as uadmin user;"
echo -e "\t${BOLD}ssh -tp 12323 uadmin@one-rpi3b.cptx86.com 'sudo ${COMMAND_NAME} /etc/docker/certs.d/daemon ca.pem'${NORMAL}\n"  # 3.550
echo -e "   Administrator checking a Docker private registry CA by including sudo.  To"
echo    "   use ${COMMAND_NAME} on a remote hosts (two-rpi3b.cptx86.com) with"
echo    "   default ssh port as uadmin user;"
echo -e "\t${BOLD}ssh -t uadmin@two-rpi3b.cptx86.com 'sudo ${COMMAND_NAME} /etc/docker/certs.d/two.cptx86.com:17315 ca.crt'${NORMAL}"  # 3.550

echo -e "\n${BOLD}SEE ALSO${NORMAL}"                                                        # 3.550
echo    "   cluster-command.sh (https://github.com/BradleyA/Linux-admin/tree/master/cluster-command)"  # 3.550

echo -e "\n${BOLD}AUTHOR${NORMAL}"                                                          # 3.550
echo    "   ${COMMAND_NAME} was written by Bradley Allen <allen.bradley@ymail.com>"         # 3.550

echo -e "\n${BOLD}REPORTING BUGS${NORMAL}"                                                  # 3.550
echo    "   Report ${COMMAND_NAME} bugs https://github.com/BradleyA/docker-security-infrastructure/issues/new"  # 3.550

###  Production standard 5.3.550 Copyright                                                  # 3.550
echo -e "\n${BOLD}COPYRIGHT${NORMAL}"                                                       # 3.550
echo    "   Copyright (c) 2020 Bradley Allen"                                               # 3.550
echo    "   MIT License https://github.com/BradleyA/docker-security-infrastructure/blob/master/LICENSE"  # 3.550

#    echo -e "\n${BOLD}HISTORY${NORMAL}"                                                    # 3.550
}

#    Date and time function ISO 8601
get_date_stamp() {
  DATE_STAMP=$(date +%Y-%m-%dT%H:%M:%S.%6N%:z)
  TEMP=$(date +%Z)
  DATE_STAMP="${DATE_STAMP} (${TEMP})"
}

#    Fully qualified domain name FQDN hostname
LOCALHOST=$(hostname -f)

#    Version
#    Assumptions for the next two lines of code:  The second line in this script includes the script path & name as the second item and
#    the script version as the third item separated with space(s).  The tool I use is called 'markit'. See example line below:
#       template/template.sh  3.517.783  2019-09-13T18:20:42.144356-05:00 (CDT)  https://github.com/BradleyA/user-files.git  uadmin  one-rpi3b.cptx86.com 3.516  
SCRIPT_NAME=$(head -2 "${0}" | awk '{printf $2}')
SCRIPT_VERSION=$(head -2 "${0}" | awk '{printf $3}')
if [[ "${SCRIPT_NAME}" == "" ]] ; then SCRIPT_NAME="${0}" ; fi
if [[ "${SCRIPT_VERSION}" == "" ]] ; then SCRIPT_VERSION="v?.?" ; fi

#    GID
GROUP_ID=$(id -g)

###  Production standard 2.3.529 log format (WHEN WHERE WHAT Version Line WHO UID:GID [TYPE] Message)
new_message() {  #  $1="${LINENO}"  $2="DEBUG INFO ERROR WARN"  $3="message"
  get_date_stamp
  echo -e "${NORMAL}${DATE_STAMP} ${LOCALHOST} ${SCRIPT_NAME}[$$] ${SCRIPT_VERSION} ${1} ${USER} ${UID}:${GROUP_ID} ${BOLD}[${2}]${NORMAL}  ${3}"
}

#    INFO
new_message "${LINENO}" "${YELLOW}INFO${WHITE}" "  Started..." 1>&2

#    Added following code because USER is not defined in crobtab jobs
if ! [[ "${USER}" == "${LOGNAME}" ]] ; then  USER=${LOGNAME} ; fi
if [[ "${DEBUG}" == "1" ]] ; then new_message "${LINENO}" "DEBUG" "  Setting USER to support crobtab...  USER >${USER}<  LOGNAME >${LOGNAME}<" 1>&2 ; fi

#    DEBUG
if [[ "${DEBUG}" == "1" ]] ; then new_message "${LINENO}" "DEBUG" "  Name_of_command >${SCRIPT_NAME}< Name_of_arg1 >${1}< Name_of_arg2 >${2}< Name_of_arg3 >${3}<  Version of bash ${BASH_VERSION}" 1>&2 ; fi

###  Production standard 9.3.513 Parse CLI options and arguments
while [[ "${#}" -gt 0 ]] ; do
  case "${1}" in
    --help|-help|help|-h|h|-\?)  display_help | more ; exit 0 ;;
    --usage|-usage|usage|-u)  display_usage ; exit 0  ;;
    --version|-version|version|-v)  echo "${SCRIPT_NAME} ${SCRIPT_VERSION}" ; exit 0  ;;
    *) break ;;
  esac
done
if [[ "${DEBUG}" == "1" ]] ; then new_message "${LINENO}" "DEBUG" "  Variable... CERT_DIR >${CERT_DIR}< CA_CERT >${CA_CERT}<" 1>&2 ; fi

###  Production standard 7.0 Default variable value
#    Order of precedence: CLI argument, environment variable, default code
if [[ $# -ge  1 ]]  ; then CERT_DIR=${1} ; elif [[ "${CERT_DIR}" == "" ]] ; then CERT_DIR=${DEFAULT_CERT_DIR} ; fi
if [[ $# -ge  2 ]]  ; then CA_CERT=${2}  ; elif [[ "${CA_CERT}" == "" ]]  ; then CA_CERT=${DEFAULT_CA_CERT}   ; fi
if [[ "${DEBUG}" == "1" ]] ; then new_message "${LINENO}" "DEBUG" "  Variable... CERT_DIR >${CERT_DIR}< CA_CERT >${CA_CERT}<" 1>&2 ; fi

#
if [[ -s "${CERT_DIR}/${CA_CERT}" ]] ; then
  #  Get certificate start and expiration date of ${CA_CERT} file
  CA_CERT_START_DATE=$(openssl x509 -in "${CERT_DIR}/${CA_CERT}" -noout -startdate | cut -d '=' -f 2)
  CA_CERT_START_DATE=$(date -d"${CA_CERT_START_DATE}" +%Y-%m-%dT%H:%M:%S-%Z)
  CA_CERT_EXPIRE_DATE=$(openssl x509 -in "${CERT_DIR}/${CA_CERT}" -noout -enddate | cut -d '=' -f 2)
  CA_CERT_EXPIRE_DATE=$(date -d"${CA_CERT_EXPIRE_DATE}" +%Y-%m-%dT%H:%M:%S-%Z)
  DAYS_VALID=$(echo "( $(date -d "${CA_CERT_EXPIRE_DATE}" +%s) - $(date -d "${CA_CERT_START_DATE}" +%s)) / (24*3600)" | bc)
#    Help hint
  echo -e "\n\tDocker CA, ${YELLOW}${CERT_DIR}/${CA_CERT}${WHITE}, is valid for ${BOLD}${YELLOW}${DAYS_VALID} days${NORMAL}"
  echo -e "\tfrom ${BOLD}${YELLOW}${CA_CERT_START_DATE}${NORMAL} to ${BOLD}${YELLOW}${CA_CERT_EXPIRE_DATE}${NORMAL}."
  echo -e "\n\tOther ${CA_CERT} files in directory ${CERT_DIR} are listed below:${CYAN}${BOLD}"
  cd "${CERT_DIR}"
  ls -1 "${CA_CERT}"* | sed 's/^/        /' 
  echo -e "${NORMAL}"
else
#    Help hint
  echo "Cannot access ${BOLD}${YELLOW}${CERT_DIR}/${CA_CERT}${NORMAL}: Permission denied or No such file or directory"
fi

#
new_message "${LINENO}" "${YELLOW}INFO${WHITE}" "  Operation finished..." 1>&2
###
